include/PetriEngine/Reachability/ReachabilitySearch.h

```diff
@@ -196,6 +216,93 @@ namespace PetriEngine {
+                                                    int depthRandomWalk, const int incRandomWalk)
+        {
+            // Set up state
+            searchstate_t ss;
+            ss.enabledTransitionsCount.resize(_net.numberOfTransitions(), 0);
+            ss.expandedStates = 0;
+            ss.exploredStates = 1;
+            ss.heurquery = queries.size() >= 2 ? std::rand() % queries.size() : 0;
+            ss.usequeries = usequeries;
+            const PQL::Condition *query = queries[ss.heurquery].get();
+
+            // Set up working area
+            _initial.setMarking(_net.makeInitialMarking());
+            Structures::State candidate; // Candidate for the next step, modified by the generator
+            Structures::State currentStepState;
+            candidate.setMarking(_net.makeInitialMarking());
+            currentStepState.setMarking(_net.makeInitialMarking());
+
+            Structures::RandomWalkStateSet states(_net, _kbound, query, seed); // State set
+            G generator = _makeSucGen<G>(_net, queries);                       // Successor generator
+
+            // Check initial marking
+            if(ss.usequeries)
+            {
+                if(checkQueries(queries, results, _initial, ss, &states))
+                {
+                    if(statisticsLevel != StatisticsLevel::None)
+                        printStats(ss, &states, statisticsLevel);
+                    _max_tokens = states.maxTokens();
+                    return true;
+                }
+            }
+
+            const int maxDepthValue = std::numeric_limits<int>::max() - incRandomWalk;
+            while(true) {
+                // Start a new random walk
+                states.newWalk();
+
+                // Search! Each turn is a random step
+                for(int stepCounter = 0; stepCounter < depthRandomWalk; ++stepCounter) {
+                    // The currentStepMarking is the nextMarking computed in the previous step
+                    if (!states.nextStep(currentStepState.marking())) {
+                        // No candidate found at the previous step, do a new walk
+                        break;
+                    }
+                    generator.prepare(&currentStepState);
+
+                    while(generator.next(candidate)) {
+                        ss.enabledTransitionsCount[generator.fired()]++;
+                        ss.exploredStates++;
+
+                        if (checkQueries(queries, results, candidate, ss, &states)) {
+                            if(statisticsLevel != StatisticsLevel::None)
+                                printStats(ss, &states, statisticsLevel);
+                            _max_tokens = states.maxTokens();
+                            _satisfyingMarking = (size_t)candidate.marking(); // This is bad, but _satisfyingMarking is only use for printing the trace (never the case for RandomWalk)
```

Comment:
Add a warning/error if the random-walk is called with traces enabled. Skip the assignment here.

Reply:
Why @petergjoel ? We would like to get the trace back from RandowWalk (if it finds it)?